### PR TITLE
use separate PRIVMSG handler for CTCPs, add_ctcp()

### PR DESCRIPTION
**Pull request by cooper**
#407

### CODE REVIEW

File: utils.py [hunk 1]
```diff
@@ -363,6 +382,55 @@ def add_cmd(self, func, name=None, featured=False, aliases=None):
...
+                try:
+                    func.is_global
+                except AttributeError:
+                    log.info('(%s/%s) CTCP command %r from %s is not global; ignored', irc.name, self.name, cmd, irc.get_hostmask(source))
+                    continue
+                    
+            # Run handler
+            try:
+                func(irc, source, self.uids.get(irc.name), args)
+            except Exception as e:
```
Comment (GLolol) [inline utils.py]: Just do `except:` since the exception value isn't used directly.

File: utils.py [hunk 2]
```diff
@@ -363,6 +382,55 @@ def add_cmd(self, func, name=None, featured=False, aliases=None):
...
+            # Fall back to global handler on PyLink service
+            funcs = world.services['pylink'].ctcp_commands[cmd]
+            need_global = True
+        else:
+            log.info('(%s/%s) Received unknown CTCP command %r from %s', irc.name, self.name, cmd, irc.get_hostmask(source))
+            return
+                
+        # Run handlers
+        log.info('(%s/%s) Handling CTCP %r for %s', irc.name, self.name, cmd, irc.get_hostmask(source))
+        for func in funcs:
```
Comment (GLolol) [inline utils.py]: Per above, I think CTCP listeners should be singular (so this iteration isn't needed).

File: plugins/ctcp.py [hunk 3]
```diff
@@ -5,25 +5,24 @@
...
 from pylinkirc.log import log
 
-def handle_ctcpversion(irc, source, args):
+@utils.add_ctcp
+def version(irc, source, target, args):
     """
     Handles CTCP version requests.
     """
-    irc.msg(source, '\x01VERSION %s\x01' % irc.version(), notice=True)
+    irc.msg(source, '\x01VERSION %s\x01' % irc.version(), notice=True, source=target)
```
Comment (cooper) [inline plugins/ctcp.py]: when I tried using `reply()`, all responses came from PyLink instead of the queried service bot

Comment (GLolol) [inline plugins/ctcp.py]: `irc.reply()` also has a `source=` argument

**Comment by GLolol** [comment]
@cooper ping

**Pull request closed without merging**